Uptimerobot: add possibbility to add and remove monitors

Ninjiner opened the pull request:
##### SUMMARY
Although this code needs improvement, it makes it possible to add and remove monitors from uptimerobot. Also it interacts with the api v2 of uptimerobot. These three essential functionalities were missing before and I think they are the main reason you would use this module.

This would resolve this two issues:
- https://github.com/ansible/ansible/issues/36984
- https://github.com/ansible/ansible/issues/56994

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
uptimerobot

##### ADDITIONAL INFORMATION
I`m open for changes and improvements.

##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request
- Docs Pull Request
- Feature Pull Request
- New Module Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

Ninjiner committed:
add possibbility to add and remove monitors

##### SUMMARY
Although this code needs improvement, it makes it possible to add and remove monitors from uptimerobot. Also it interacts with the api v2 of uptimerobot. These three essential functionalities were missing before and I think they are the main reason you would use this module.

This would resolve this two issues:
- https://github.com/ansible/ansible/issues/36984
- https://github.com/ansible/ansible/issues/56994

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
uptimerobot

##### ADDITIONAL INFORMATION
I`m open for changes and improvements.

Ninjiner committed:
fix shippable indicated errors

- Fix missing version_added tag in documentation
- Fix invalid pep8 comment

Akasurde commented on lib/ansible/modules/monitoring/uptimerobot.py:
```diff
@@ -6,19 +6,12 @@
 __metaclass__ = type
 
-
-ANSIBLE_METADATA = {'metadata_version': '1.1',
```
Please don't remove ANSIBLE_METADATA, it is required.

Ninjiner replied:
I am sorry, I just moved it to:
https://github.com/Ninjiner/ansible/blob/patch-1/lib/ansible/modules/monitoring/uptimerobot.py#L73
@Akasurde is that an issue? Shall i move it back to the top?

Akasurde replied:
`ANSIBLE_METADATA` is kept after `__metaclass__` for all modules so you might want to keep the way it was.

Ninjiner replied:
I moved it back, where it was.

Akasurde commented on lib/ansible/modules/monitoring/uptimerobot.py:
```diff
@@ -6,19 +6,12 @@
-    - This module will let you start and pause Uptime Robot Monitoring
-author: "Nate Kingsley (@nate-kingsley)"
+    - This module lets you create, remove, pause and start monitors in Uptimerobot
+author: "Ninjiner (@Ninjiner)"
```
```suggestion
author:
- "Nate Kingsley (@nate-kingsley)"
- "Ninjiner (@Ninjiner)"
```

Akasurde commented on lib/ansible/modules/monitoring/uptimerobot.py:
```diff
@@ -27,124 +20,168 @@
         description:
             - Define whether or not the monitor should be running or paused.
```
Add a note stating that `created` and `absent` are added in version 2.10

Also, `created` can be alias to `present` to match other modules.

Ninjiner replied:
https://github.com/Ninjiner/ansible/commit/46def8f3ee4dd6464e5cc4044e81c7b8555748f9

Ninjiner replied:
https://github.com/ansible/ansible/pull/63843/commits/36dc86bec6550e742aa5a4cb6e49bd54b72e067f

Akasurde commented on lib/ansible/modules/monitoring/uptimerobot.py:
```diff
@@ -27,124 +20,168 @@
+        choices: [ "started", "paused", "created", "absent" ]
+    url:
+        description:
+            - Url to be checked
```
```suggestion
            - URL to be checked.
```

Akasurde commented on lib/ansible/modules/monitoring/uptimerobot.py:
```diff
@@ -27,124 +20,168 @@
+        version_added: '2.10'
+    name:
+        description:
+            - The friendly name of the monitor
```
```suggestion
            - The friendly name of the monitor.
```

Akasurde commented on lib/ansible/modules/monitoring/uptimerobot.py:
```diff
@@ -27,124 +20,168 @@
+            state=dict(required=True, choices=['started', 'paused', 'absent', 'created']),
             apikey=dict(required=True, no_log=True),
-            monitorid=dict(required=True)
+            monitorid=dict(required=False)
```
Add a note in documentation stating that `monitorid` is not required parameter from 2.10 and a porting guide note [here](https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/porting_guides/porting_guide_2.10.rst#noteworthy-module-changes)

Ninjiner replied:
https://github.com/Ninjiner/ansible/commit/f2c2822c08a7bac19e0a0d111e63cc709e801b97

Ninjiner replied:
https://github.com/Ninjiner/ansible/commit/36e2719a1a82e24c04c34dc03e4115c2c8903da9

Akasurde commented on lib/ansible/modules/monitoring/uptimerobot.py:
```diff
@@ -27,124 +20,168 @@
+        data['api_key'] = self.params['api_key']
+        data = json.dumps(data)
+        self.uri = API_BASE + "getMonitors"
+        req = Request()
```
Make a common method in class to get url and return json loads response.

Ninjiner replied:
Is this what you had in mind: https://github.com/ansible/ansible/pull/63843/commits/0ae4174aab4a9df71b6cf918f86e53b70a0ce54c ?

Akasurde commented on lib/ansible/modules/monitoring/uptimerobot.py:
```diff
@@ -27,124 +20,168 @@
+        req = Request()
+        state = req.post(url=self.uri, data=data, headers=self.headers)
+        state = state.read()
+        state = json.loads(state)
```
```suggestion
        state = json.loads(to_text(state.read(), errors='surrogate_or_strict'))
```

This makes this call future proof.

Ninjiner replied:
https://github.com/Ninjiner/ansible/commit/febdd0a55d5ccab79aa34687e9cb4c91d4889c08

Akasurde reviewed (changes requested):
Thanks for the contribution.

Ninjiner committed:
accept review of Akasurde

Add changes proposed by Akasurde

Co-Authored-By: Abhijeet Kasurde <akasurde@redhat.com>

Ninjiner committed:
fix url description

Co-Authored-By: Abhijeet Kasurde <akasurde@redhat.com>

Ninjiner committed:
fix monitor name description

Co-Authored-By: Abhijeet Kasurde <akasurde@redhat.com>

Ninjiner committed:
make state future proof

Co-Authored-By: Abhijeet Kasurde <akasurde@redhat.com>

Ninjiner commented:
@Akasurde here: 
https://github.com/Ninjiner/ansible/blob/patch-1/lib/ansible/modules/monitoring/uptimerobot.py#L120

Did you mean the function "to_text"? Because as it sais in the tests, there is no var with that name.

Ninjiner committed:
fix monitorid requirement in doc

State in the documentation, that it is not a required parameter (propose by @Akasurde)

Ninjiner committed:
fix monitorid version_added in doc

Add the version added note for the monitorid (propose by @Akasurde)

Ninjiner committed:
fix state version_added in doc

Add the version added note for the state

Ninjiner committed:
convert state to unicode

- ensure that state is unicode
- shorten the process of getting the state in unicode

Ninjiner commented:
> The test `ansible-test sanity --test validate-modules` [[explain](https://docs.ansible.com/ansible/devel/dev_guide/testing/sanity/validate-modules.html)] failed with 2 errors:
> 
> ```
> lib/ansible/modules/monitoring/uptimerobot.py:0:0: option-incorrect-version-added: version_added for new option (monitorid) should be None. Currently '2.10'
> lib/ansible/modules/monitoring/uptimerobot.py:0:0: option-incorrect-version-added: version_added for new option (state) should be None. Currently '2.10'
> ```
> 
> The test `ansible-test sanity --test pylint` [[explain](https://docs.ansible.com/ansible/devel/dev_guide/testing/sanity/pylint.html)] failed with 1 error:
> 
> ```
> lib/ansible/modules/monitoring/uptimerobot.py:123:27: undefined-variable: Undefined variable 'to_text'
> ```
> 
> [click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

@Akasurde why does it complain about the version_added tags?

Ninjiner committed:
fix version_added and add present state

as proposed by Abhijeet Kasurde <akasurde@redhat.com> in https://github.com/ansible/ansible/pull/63843#discussion_r338879645

Ninjiner committed:
move metadata block back

move the metadata block back to the place it initially was

Ninjiner committed:
add generic api request function

implement a function which gets data from an api and returns json dump

Ninjiner commented:
@Akasurde I think, I changed everything as requested. Is there anything else I need to change?

Akasurde commented on lib/ansible/modules/monitoring/uptimerobot.py:
```diff
@@ -27,124 +27,167 @@
-    return jsonresult['stat']
+class UptimeRobot:
+    def __init__(self, params, api_methods):
+        self.headers = {}
```
```suggestion
        self.headers = {
        	'Content-Type': "application/json"
        }
```

Ninjiner replied:
Fixed in https://github.com/Ninjiner/ansible/commit/2398716806690f15d965275149022a58b4f4c8df

Akasurde commented on lib/ansible/modules/monitoring/uptimerobot.py:
```diff
@@ -27,124 +27,167 @@
+        self.params = self.check_dict(self.params)
+        for k, v in self.params.items():
+            if k == 'type':
+                if v == 'dns':
```
```suggestion
                self.params[k] = 2 if v == 'dns' else 1
```

Ninjiner replied:
Fixed in https://github.com/Ninjiner/ansible/commit/2165e03dd4ebc6e6b48104445526f2cfce879cf8

Akasurde commented on lib/ansible/modules/monitoring/uptimerobot.py:
```diff
@@ -27,124 +27,167 @@
+        self.api_methods = api_methods
+        self.params = params
+        self.params = self.check_dict(self.params)
+        for k, v in self.params.items():
```
This can be moved to `check_dict` method.

Ninjiner replied:
I'm sorry, but what do you want to move to the `check_dict` function?
Because the first three lines you marked should be in the constructor.

Ninjiner replied:
@Akasurde
- The `self.api_methods` variable is used multiple times in the constructor, so it must stay there.
- I will shorten the generation of `self.params` by directly passing params to `check_dict`
- The loop I will move to the function itself

Akasurde commented on lib/ansible/modules/monitoring/uptimerobot.py:
```diff
@@ -27,124 +27,167 @@
+    result['message'] = state['stat']
+    result['original_message'] = state
+    if result['message'] != 'ok':
+        module.fail_json(msg='Could not perform action {0}'.format(api_method))
```
I am not able to create a new monitor -

```
fatal: [localhost]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "apikey": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "check_type": null,
            "monitorid": null,
            "name": "myMonitor_002",
            "state": "present",
            "url": "http://www.my-domain.com"
        }
    },
    "msg": "Could not perform action newMonitor"
}
``

Ninjiner replied:
Interesting, I can create and delete.
I use a role with one task like this:
```
---
- name: create Monitor
  custom_uptimerobot:
    state: absent
    url: https://www.google.ch
    name: TEST
    apikey: "MY_KEY"
  delegate_to: localhost
...
```
And a Playbook like this:
```
---
- hosts:
  - somehost
  roles:
    - uptimerobot
  gather_facts: no
...
```

Ninjiner replied:
@Akasurde Try if it works if you remove the null values. The `check_type` and `monitorid` keys can be ignored because they are optional.

Ninjiner committed:
simplify header initialisation

Ninjiner committed:
simplify declaration of check type

drzraf commented:
Nice opportunity to add `check_type = keyword`?

drzraf commented on lib/ansible/modules/monitoring/uptimerobot.py:
```diff
@@ -27,124 +27,165 @@
-            state=dict(required=True, choices=['started', 'paused']),
+            name=dict(required=True),
+            url=dict(required=True),
+            check_type=dict(required=False, choices=['http', 'dns']),
```
`dns`?
According to documentation: https://uptimerobot.com/api

### monitor>type ###
1 - HTTP(s)
2 - Keyword
3 - Ping
4 - Port

Ninjiner replied:
Yeah, my fault, thank you for telling me.

Ninjiner replied:
https://github.com/ansible/ansible/pull/63843/commits/0b0736f71caabec059c88e72dc975d811d5ba5b3

drzraf commented:
For reference, an alternative module implementation based on v2 API lives here : https://github.com/ansible/ansible/pull/63843

But it does not seems to support check-mode either (which is one of the biggest drawback of all implementations)

Commit:
refactor check_dict method

Ninjiner committed:
fix usage of non existing monitor type

As noticed by drzraf 'dns' is not a valid option to create a moonitor.
In this commit this is changed to the valid option ping (3).

Ninjiner commented:
> Nice opportunity to add `check_type = keyword`?
> I suggest making `check_type` string a `type=dict, options=...` (named simply `check`) having in its minimal form: `{type: HTTP(s)|ping|port|Keyword}` (Corresponding to 1-4 values of the API)
> 
> It could be easily extended with all type-related params:
> Eg: `{type: "port", sub_type: SMTP}` or `{type: "Keyword", keyword_type: ..., keyword_value: ...}`



> Nice opportunity to add `check_type = keyword`?
> I suggest making `check_type` string a `type=dict, options=...` (named simply `check`) having in its minimal form: `{type: HTTP(s)|ping|port|Keyword}` (Corresponding to 1-4 values of the API)
> 
> It could be easily extended with all type-related params:
> Eg: `{type: "port", sub_type: SMTP}` or `{type: "Keyword", keyword_type: ..., keyword_value: ...}`

I am total with you on this. The reason I did not do this is simply, that then your input is not validated. So the user is responsible to enter something that is valid. Otherwise, I would (in the future), expand this, so you can use any type.

Ninjiner commented:
> For reference, an alternative module implementation based on v2 API lives here : #63843 from @felipesantiago
> 
> But it does not seem to support check-mode either (which is one of the biggest drawback of all implementations when it comes to provisioning 3rd-party services)

Is this supposed to be a reference to this PR?

drzraf commented:
> I am total with you on this. The reason I did not do this is simply, that then your input is not validated. So the user is responsible to enter something that is valid. Otherwise, I would (in the future), expand this, so you can use any type.

I'm pretty sure supported keyword (even if not completely validated) is better than not having it at all.

You can try processing the spec' in two steps.
1. Step one, `check_type` is validated against a `dict(type=str)` and `check_opts` as an opaque `dict`.
2. Once `check_type` got validated, update the validation spec' of `check_opts` to match the check type and run the validation again.

I tried such a thing in #40053:
https://github.com/ansible/ansible/blob/dca8cf11634045ca774a3d2dd918e7f5da34dc96/lib/ansible/modules/source_control/gitlab_service.py#L193-L195

https://github.com/ansible/ansible/blob/dca8cf11634045ca774a3d2dd918e7f5da34dc96/lib/ansible/modules/source_control/gitlab_service.py#L372-L374

Ninjiner commented:
@drzraf unfortunately at the moment i have almost no spare time to implement this. Would you like to help me out here?
